src/liquidations.ts

```diff
@@ -40,20 +40,129 @@ export function isBadDebt(user: PositionsEstimate): boolean {
+    0,
+    estimate.totalEffectiveLiabilities * 1.06 - estimate.totalEffectiveCollateral
   );
+  let liqPercent = calculateLiqPercent(auctionEstimate, liabilitesToReduce);
+  while (liqPercent > 100 || liqPercent === 0) {
+    if (liqPercent === 0) {
+      let nextLiability = effectiveLiabilities.shift();
+      if (nextLiability === undefined) {
+        // No more collaterals to liquidate
+        let nextCollateral = effectiveCollaterals.shift();
+        if (nextCollateral === undefined) {
+          return {
+            auctionPercent: 100,
+            lot: Array.from(auction.collateral).map(([index]) => pool.metadata.reserveList[index]),
+            bid: Array.from(auction.liabilities).map(([index]) => pool.metadata.reserveList[index]),
+          };
+        }
+        auction.collateral.set(nextCollateral[0], user.collateral.get(nextCollateral[0])!);
+      } else {
+        auction.liabilities.set(nextLiability[0], user.liabilities.get(nextLiability[0])!);
+      }
+    } else if (liqPercent >= 101) {
+      let nextCollateral = effectiveCollaterals.shift();
+      if (nextCollateral === undefined) {
+        // No more collaterals to liquidate
+        return {
+          auctionPercent: liqPercent,
+          lot: Array.from(auction.collateral).map(([index]) => pool.metadata.reserveList[index]),
+          bid: Array.from(auction.liabilities).map(([index]) => pool.metadata.reserveList[index]),
+        };
+      }
+      auction.collateral.set(nextCollateral[0], user.collateral.get(nextCollateral[0])!);
+    }
+    auctionEstimate = PositionsEstimate.build(pool, oracle, auction);
+    liqPercent = calculateLiqPercent(auctionEstimate, liabilitesToReduce);
+  }
+
+  return {
+    auctionPercent: liqPercent,
+    lot: Array.from(auction.collateral).map(([index]) => pool.metadata.reserveList[index]),
+    bid: Array.from(auction.liabilities).map(([index]) => pool.metadata.reserveList[index]),
+  };
+}
+
+function calculateLiqPercent(positions: PositionsEstimate, liabilitesToReduce: number) {
+  let avgCF = positions.totalEffectiveCollateral / positions.totalSupplied;
+  let avgLF = positions.totalEffectiveLiabilities / positions.totalBorrowed;
+  let estIncentive = 1 + (1 - avgCF / avgLF) / 2;
+  // The factor by which we can reduce the liabilities
+  let liabilityReductionFactor = avgLF * 1.06 - avgCF * estIncentive;
+  let totalRemoveableLiabilities = liabilityReductionFactor * positions.totalBorrowed;
+
+  if (totalRemoveableLiabilities < liabilitesToReduce) {
+    return 0; // Not enough liabilites or a high enough reduction factor to return position to desired health factor
+  }
+
+  let liqPercent = Math.min(
+    Math.round((liabilitesToReduce / totalRemoveableLiabilities) * 100),
+    100
+  );
+
+  let requiredRawCollateral = (liqPercent / 100) * positions.totalBorrowed * estIncentive;
+
+  if (requiredRawCollateral > positions.totalSupplied) {
+    return 101; // Not enough collateral to cover the liquidation
+  }
```

Comment:
IMO, these percentages should probably be flipped.

Liq percent > 100 means we need to liquidate more (add more liabilities), and 0 implies the liquidation would not process (not enough collateral).

Reply:
FWIW this would allow you to remove the special case for `totalRemoveableLiabilities < liabilitesToReduce`. You could just return liq percent if `requiredRawCollateral > positions.totalSupplied` and you remove the `Math.min` cap of 100.